nemo/collections/nlp/models/information_retrieval/bert_dpr_model.py

```diff
@@ -0,0 +1,277 @@
+        scores = torch.cat(
+            (torch.matmul(q_vectors, p_positives.T), torch.einsum("ij,ipj->ip", q_vectors, p_negatives)), dim=1
+        )
+
+        return scores
+
+    def training_step(self, batch, batch_idx):
+        """
+        Lightning calls this inside the training loop with the data from the training dataloader
+        passed in as `batch`.
+        """
+        # forward pass
+        q_input_ids, q_input_mask, q_input_type_ids, p_input_ids, p_input_mask, p_input_type_ids = batch
+        batch_size, num_passages, p_seq_length = p_input_ids.size()
+        q_seq_length = q_input_ids.size()[-1]
+
+        scores = self(
+            q_input_ids=q_input_ids.view(-1, q_seq_length),
+            q_token_type_ids=q_input_type_ids.view(-1, q_seq_length),
+            q_attention_mask=q_input_mask.view(-1, q_seq_length),
+            p_input_ids=p_input_ids.view(-1, p_seq_length),
+            p_token_type_ids=p_input_type_ids.view(-1, p_seq_length),
+            p_attention_mask=p_input_mask.view(-1, p_seq_length),
+        ).view(batch_size, 1, batch_size + num_passages - 1)
+        scores = torch.log_softmax(scores, dim=-1)
+
+        labels = torch.arange(batch_size)[:, None].long().to(scores.device)
+        train_loss = self.loss(logits=scores, labels=labels, output_mask=torch.ones_like(labels))
+
+        tensorboard_logs = {'train_loss': train_loss, 'lr': self._optimizer.param_groups[0]['lr']}
+        return {'loss': train_loss, 'log': tensorboard_logs}
+
+    def validation_step(self, batch, batch_idx):
+        """
+        Lightning calls this inside the validation loop with the data from the validation dataloader
+        passed in as `batch`.
+        """
+        (
+            q_input_ids,
+            q_input_mask,
+            q_input_type_ids,
+            p_input_ids,
+            p_input_mask,
+            p_input_type_ids,
+            query_id,
+            passage_ids,
+        ) = batch
+        batch_size, num_passages, p_seq_length = p_input_ids.size()
+        q_seq_length = q_input_ids.size()[-1]
+
+        scores = self(
+            q_input_ids=q_input_ids.view(-1, q_seq_length),
+            q_token_type_ids=q_input_type_ids.view(-1, q_seq_length),
+            q_attention_mask=q_input_mask.view(-1, q_seq_length),
+            p_input_ids=p_input_ids.view(-1, p_seq_length),
+            p_token_type_ids=p_input_type_ids.view(-1, p_seq_length),
+            p_attention_mask=p_input_mask.view(-1, p_seq_length),
+        ).view(batch_size, 1, batch_size + num_passages - 1)
+
+        data_for_val = {
+            "scores": scores[0, 0, :].detach().cpu().numpy(),
+            "query_id": query_id.item(),
+            "passage_ids": passage_ids[0].detach().cpu().numpy(),
+        }
+        return data_for_val
+
+    def validation_epoch_end(self, outputs):
+        """
+        Called at the end of validation to aggregate outputs.
```

Comment:
where are you all gathering outputs from all gpus?
if you dont do that, you need to use PLT.metrics which does all reduce sum/mean